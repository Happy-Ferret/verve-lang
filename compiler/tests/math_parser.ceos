(defn parseFactorR (sum expr)
  (if (gt (count expr) 0)
      ((lambda (c)
        (if (and
            (gte c '0')
            (lte c '9'))
          (parseFactorR (add (mul 10 sum) (sub c '0')) (substr expr 1))
          (list sum expr)
        )
      ) (at expr 0))
      (list sum expr)
  )
)

(defn parseFactor (e)
  (parseFactorR 0 e)
)

(defn evalExpr (e)
  ((lambda (list)
    ((lambda (left expr)
      (if (gt (count expr) 0)
        ((lambda (op expr)
          (if (equals op '+')
            (add left (evalExpr expr))
            (if (equals op '-')
              (sub left (evalExpr expr))
            )
          )
        ) (at expr 0) (substr expr 1))
        left
      )
    ) (at list 0) (at list 1))
  ) (parseFactor e))
)

(print (evalExpr "22"))
(print (evalExpr "22+22"))
(print (evalExpr "22-11"))
(print (evalExpr "10+51-17"))
