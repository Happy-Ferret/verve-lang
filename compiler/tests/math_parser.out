22
44
11
44
